### ISSUE TITLE
Unable to SYNC test devices in Google Home companion app

### ISSUE THREAD

**Issue by pantongren** [actions-on-google/actionssdk-smart-home-nodejs#26]
After I choose the "device" and login there's a progressdialog with "Linking your ... account".. then I got an error: "Couldn't update the settings. Check your connection",why?

**Comment by Fleker** [comment]
Can you try a few times and see if it works after another check?

**Comment by pantongren** [comment]
@Fleker I have tried many times, 
I was in accordance with the README step by step .When I  copy the url "https://xxxxx..ngrok.io/" ,open it in a browser and clicked on the submit button to login and then it jumped to a blank page, nothing at all. Next,I try it in Google Home App ,and it shows"Couldn't update the settings. Check your connection".

**Comment by Fleker** [comment]
Is your local server getting the authentication request and the SYNC event?

**Comment by pantongren** [comment]
@Fleker ,yes,here is my log:
getUser rick
logging in  { uid: '1234',
  name: 'rick',
  password: 'oldman',
  tokens: [ 'psokmCxKjfhk7qHLeYd1' ] }
login successful  rick
authCode successful  cuvp84qffp4w84s04cs8w0o8o
POST /login 302 1.865 ms - 1478
/token query {}
/token body { grant_type: 'authorization_code',
  code: 'cuvp84qffp4w84s04cs8w0o8o',
  redirect_uri: 'https://oauth-redirect.googleusercontent.com/r/smartliving-8f0d
9',
  client_id: 'ZxjqWpsYj3',
  client_secret: 'hIMH3uWlMVrqa7FAbKLBoNUMCyLCtv' }
getClient ZxjqWpsYj3, hIMH3uWlMVrqa7FAbKLBoNUMCyLCtv
return getClient { clientId: 'ZxjqWpsYj3',
  clientSecret: 'hIMH3uWlMVrqa7FAbKLBoNUMCyLCtv' }
client { clientId: 'ZxjqWpsYj3',
  clientSecret: 'hIMH3uWlMVrqa7FAbKLBoNUMCyLCtv' }
handleAuthCode {}
getClient ZxjqWpsYj3, hIMH3uWlMVrqa7FAbKLBoNUMCyLCtv
return getClient { clientId: 'ZxjqWpsYj3',
  clientSecret: 'hIMH3uWlMVrqa7FAbKLBoNUMCyLCtv' }
getAccessToken =  { uid: '1234',
  accessToken: 'psokmCxKjfhk7qHLeYd1',
  refreshToken: 'psokmCxKjfhk7qHLeYd1',
  userId: '1234' }
return getAccessToken =  { token_type: 'bearer',
  access_token: 'psokmCxKjfhk7qHLeYd1',
  refresh_token: 'psokmCxKjfhk7qHLeYd1' }
respond success { token_type: 'bearer',
  access_token: 'psokmCxKjfhk7qHLeYd1',
  refresh_token: 'psokmCxKjfhk7qHLeYd1' }
POST /token 200 5.177 ms - 100
post /ha { 'content-type': 'application/json',
  'google-assistant-api-version': 'v1',
  authorization: 'Bearer psokmCxKjfhk7qHLeYd1',
  host: '6fe99024.ngrok.io',
  'content-length': '79',
  'user-agent': 'Mozilla/5.0 (compatible; Google-Cloud-Functions/2.1; +http://ww
w.google.com/bot.html)',
  'accept-encoding': 'gzip,deflate,br',
  'x-forwarded-proto': 'https',
  'x-forwarded-for': '66.249.84.202' }
post /ha { inputs: [ { intent: 'action.devices.SYNC' } ],
  requestId: '6345068453055187986' }
post /ha SYNC
query { uid: '1234',
  auth: 'psokmCxKjfhk7qHLeYd1',
  requestId: '6345068453055187986' }
POST /ha 200 1.806 ms - 146

**Comment by pantongren** [comment]
@Fleker   What's wrong with my log, sir?

**Comment by Fleker** [comment]
Good evening, I've been out of office.
Your logs suggest no issue. Your server is getting all the events. Your
phone and the home app dont appear to be getting the response, either due
to a problem in the app or some sort of timeout in the app. I'm unsure of a
good way to fix at the moment. I'm going to raise the question internally
in the morning.

Have a good night!

**Comment by marziman** [comment]
@Fleker +1 for me. I also get this error, after trying to link the account and it seems the browser cant recall the app.

**Comment by pantongren** [comment]
@marziman Have you solved this problem?

**Comment by pantongren** [comment]
@Fleker thank you sir, waiting for  your good news.

**Comment by bartvollebregt** [comment]
+1 I'm having the exact same issue.

**Comment by mwidlok** [comment]
@marziman Hi, 
when I'm using Google Home App on iOS, I'm able to display login form within my system when using Home Control -> Add devices. 
Then after successful login I get "Something went wrong. Try Again"
No SYNC call received in API. 
Do you have the same problem?

**Comment by marziman** [comment]
@pantongren , @mwidlok ,

I get the error like the first comment: "Couldn't update the settings. Check your connection"
and the linking wont happen, so there is no SYNC sent.
So I think this is a Google platform issue and must be either fixed in app or platform.

I am waiting for feedback from @Fleker . I am sure he is busy checking that issue.
So lets wait for his feedback.

BR

**Comment by pantongren** [comment]
@marziman I solved this issue yesterday, we have to give a correct-format value to the constant "Data" in the datastore.js file, The original value is empty. If you run the program before you assign a vaule to the constant, the program will reports an exception and then you will get the message"Couldn't update the settings. Check your connection" on your phone, if you give a uncorrect-format value to the constant, the output logs are nomal, but you still get the message"Couldn't update the settings. Check your connection" on your phone.

**Comment by mwidlok** [comment]
@Fleker Any updates on this issue?
Our client are going to Berlin IFA show next week, to present their Smart Home system. He wants to show Google Home assistant features but at the moment this is impossible.

**Comment by marziman** [comment]
@pantongren,

not sure if I understand it correctly. I use a real smarthome backend with oAuth capabilities and not the mock implementation here with datastore.js.
Can you explain how you fixed it in detail? Or did you just found a quick fix for your problem..

**Comment by pantongren** [comment]
@marziman  I  didn't use a real smarthome backend with oAuth capabilities. So, I'm not sure whether it is authentication or SYNC leads to your “Couldn't update the settings. Check your connection”，You can check your server output logs after you click on the login button. if your server get the request of SYNC, I think it is the problem with your response data。The reason I failed to link my account most likely was that my server did not reply the correct-format  data  to the phone after SYNC request.  The correct-format response data like here:  {
                "id": "g:3",
                "type": "action.devices.types.SWITCH",
                "traits": [
                    "action.devices.traits.OnOff"
                ],
                "name": {
                    "defaultNames": [
                        "HAM Group"
                    ],
                    "name": "Park Lights"
                },
                "willReportState": false,
                "deviceInfo": {
                    "manufacturer": "HAM Systems",
                    "model": "HAM Group",
                    "hwVersion": "1.0"
                }
            },

**Comment by rinehaidvon** [comment]
@Fleker I also have this issue.I try to add devices through the android phone, most of the cases have failed, only a few success. From the log analysis, my server did not receive the sync call. I can be sure that my server is working well and I try to use the postman to sending sync, query and execute requests that all get the right response.

Here is my log:

login successful  rick
authCode successful  81bvf0thxsc000000000000000
GET /oauth?response_type=code&client_id=ZxjqWpsYj3&redirect_uri=https://oauth-re
direct.googleusercontent.com/r/test-smart-home&scope=email&state=CrwCQUZEXzV0bTR
TR09NQXhXcWtTc185OW9pNGFKRXpjX2xmRHVubzdBLTJ4dXJhdXVRUWs0MkVSMGVuLTd4Y181dmpDWll
sdG9MZmVVVjd0cm1lamtaY1BId2lrd3YwSWQ0X3Y0Yk4yNmpCVFJDNWJ3Y1B6V0ViLXJTYUo5dTY4NVN
YR0JJc09oMHB3ZlFuVWI4M2VhMHUyMXQxY3NfY2Y3Skx6b3VIWS1Nb3pEVXRzcG5JQkx4YVVFSnJZNWt
FQXFkY0VVX20tdlM1YnBVR1YyS01Xbk45UzdxM044MDdTRXpqTkV2TXJJdzV2VlNkTDRvZEtRN2s4YVN
iVWxEcGVyOVBoQU4tREpJRlFaTUVXWFR0UmdYSXNPWWhfYU1WQno3ZHBBSjQ3bjU2WjRzTnJpd3otaFF
KcE1ScnlfVBIcYXNzaXN0YW50LWhhbmRvZmY6Ly9jb21wbGV0ZSI-aHR0cHM6Ly9vYXV0aC1yZWRpcmV
jdC5nb29nbGV1c2VyY29udGVudC5jb20vci90ZXN0LXNtYXJ0LWhvbWUqFXJpbmVoYWlkdm9uQGdtYWl
sLmNvbTITdGVzdC1zbWFydC1ob21lX2Rldg 302 5613.128 ms - 1474
/token query {}
/token body { grant_type: 'authorization_code',
  code: '81bvf0thxsc000000000000000',
  redirect_uri: 'https://oauth-redirect.googleusercontent.com/r/test-smart-home'
,
  client_id: 'ZxjqWpsYj3',
  client_secret: 'hIMH3uWlMVrqa7FAbKLBoNUMCyLCtv' }
getClient ZxjqWpsYj3, hIMH3uWlMVrqa7FAbKLBoNUMCyLCtv
return getClient { clientId: 'ZxjqWpsYj3',
  clientSecret: 'hIMH3uWlMVrqa7FAbKLBoNUMCyLCtv' }
client { clientId: 'ZxjqWpsYj3',
  clientSecret: 'hIMH3uWlMVrqa7FAbKLBoNUMCyLCtv' }
handleAuthCode {}
getClient ZxjqWpsYj3, hIMH3uWlMVrqa7FAbKLBoNUMCyLCtv
return getClient { clientId: 'ZxjqWpsYj3',
  clientSecret: 'hIMH3uWlMVrqa7FAbKLBoNUMCyLCtv' }
getAccessToken =  { uid: '1234',
  accessToken: 'psokmCxKjfhk7qHLeYd1',
  refreshToken: 'psokmCxKjfhk7qHLeYd1',
  userId: '1234' }
return getAccessToken =  { token_type: 'bearer',
  access_token: 'psokmCxKjfhk7qHLeYd1',
  refresh_token: 'psokmCxKjfhk7qHLeYd1' }
respond success { token_type: 'bearer',
  access_token: 'psokmCxKjfhk7qHLeYd1',
  refresh_token: 'psokmCxKjfhk7qHLeYd1' }
POST /token 200 26.913 ms - 100

**Comment by Fleker** [comment]
@rinehaidvon can you post or [email me](mailto:fleker@google.com) your project-id and the account you used to create this project?

**Comment by rinehaidvon** [comment]
@Fleker 
i have already reply you, please check your email, thank you.

**Comment by BenjaminBertran** [comment]
Following the tutorial, it lead us to the same issue than @rinehaidvon, Oauth calls seem to be ok, no call to SYNC... Also the same "Couldn't update the setting. Check your connection" message in the mobile app after sign in.

Also tried to swith to our corporate sign in pages: everything seemed to be ok but led to the same result...

Don't known if it's relevant: from the mobile, just after sign in, I can see a "[Upload] response code: 500" in the mobile logs using logcat...

**Comment by Fleker** [comment]
@BenjaminBertran that may actually be pretty relevant. Can you obtain a more in-depth bug report for your device, then [email me](mailto:fleker@google.com) the zip? Are there any other logs in the app that indicate what stage the error is occurring?

**Comment by rinehaidvon** [comment]
@Fleker is there any update about this issue?

**Comment by Fleker** [comment]
Sorry, the issue seems to be affecting a select number of accounts but we
don't have an update right now.

**Comment by Fleker** [comment]
If you check the ADB logs when you link your account, do you see a 500 error when you see the sync failure in the app and a stacktrace?

`W/S3NetworkUtils: [Upload] response code: 500`

**Comment by mwidlok** [comment]
I'm using iOS app. What is the way to debug it?

Pozdrawiam,
Maciej Widłok.

2017-08-28 19:59 GMT+02:00 Nick <notifications@github.com>:

**Comment by Fleker** [comment]
I don't know how to get logs from iOS. 🤔 

This might only be available to check on Android right now.

**Comment by mwidlok** [comment]
@Fleker Could you be so kind and tell does submitting app to review and then publishing, will allow to test the app without any problems? My customer is going to IFA Exhibition in Berlin this week and wanted to show Google Home assistant features working with their smart home system.

**Comment by Fleker** [comment]
I can't guarantee that it would be working if it's published, if it passes review. We can discuss [over email](mailto:fleker@google.com) some potential workarounds for your deadline.

**Comment by mwidlok** [comment]
@Fleker Thx, right now I'm trying to setup our google home api in production env. Tomorrow when all will be set will email you to discuss workarounds.

**Comment by Fleker** [comment]
@AlexSneedMiller on Android? Can you grab a bug report after doing this in both cases?

**Comment by dbrammertz** [comment]
Having the exact same issuse with the excat same timeline. It does not seem to be a linking issuse as I got from the logcat and also from the oauth playground everything with my oauth implementation (that works also for other clients) works fine.

Nevertheless I got no sync request and real devices and the simulator keeps stating that "The settings cannot be updated. Please check your network connection". I use the exact same user (owner) for the setup as for the test.

I posted my problem on stackoverflow, google plus and opened a support ticket. Hope that this one is more promising.

Please find below my complete logcat output.

[log.txt](https://github.com/actions-on-google/actionssdk-smart-home-nodejs/files/1259193/log.txt)

**Comment by dbrammertz** [comment]
I have the exact same problem with the exact same timeline. It does not seem to be a oAuth problem. The oAuth implementation is working for other smart home assistents and also for the google oAuth playground. I grepped from logcat the following lines:

> 08-28 21:27:38.114: I/ActivityManager(1496): START u0 {dat=assistant-handoff://complete/?result_code=SUCCESS&result_message=Accounts+now+linked. cmp=com.google.android.googlequicksearchbox/com.google.android.apps.gsa.assistant.handoff.BrowserControlActivity} from uid 10030

Which basically indicates that the oAuth is succesfully completed and the google assistent has to do the next step.

I posted that problem on stackoverflow, google plus and opend also a support ticket. No reponses except other people stating that they have exactly the same problem.

I added the complete logcat output. Any help is highly appreciated!!

[log.txt](https://github.com/actions-on-google/actionssdk-smart-home-nodejs/files/1259198/log.txt)

**Comment by pevecyan** [comment]
@dbrammertz  Have you implemented Implicit flow or Authorization code flow? We had problems with first one, but Authorization code works as it should.

**Comment by dbrammertz** [comment]
We are using the Authorization Code flow. Works perfectly on https://developers.google.com/oauthplayground/, but fails for google action. Unfortunately there is no hint what is wrong. On our server I can see that a valid access_code is generated:

>2017-08-28 21:27:38  INFO  [  /  ] LoggingFilter:132 -Request: request id=173; method=POST; content type=application/x-www-form-urlencoded; uri=/dcloud-main-webapp/oauth/token; payload=
>2017-08-28 21:27:38 INFO  [  /  ] LoggingFilter:149 -Response: request id=173; payload={"access_token":"af3df1bb-0a8a-4533-8dfa-af6694dedfe0","token_type":"bearer","refresh_token":"692ee03c-475c-4c5b-a606-fb847be9386a","expires_in":1199,"scope":"auth"}

On the mobile device I can see via logcat that the access_code was accepted:
> 08-28 21:27:38.114: I/ActivityManager(1496): START u0 {dat=assistant-handoff://complete/?result_code=SUCCESS&result_message=Accounts+now+linked. cmp=com.google.android.googlequicksearchbox/com.google.android.apps.gsa.assistant.handoff.BrowserControlActivity} from uid 10030

But still the account linking is failing on the device:
<img width="224" alt="gaction errors" src="https://user-images.githubusercontent.com/6271314/29840251-9ff527d0-8d01-11e7-9075-24223c49fe22.png">

**Comment by marziman** [comment]
Same for me, there must be an issue with the app. I always get this message as reported before.

**Comment by Fleker** [comment]
Yes it does not appear to be an issue with auth. Something unusual seems to happen that prevents the `SYNC` from being sent to your smart home app. Still investigating.

**Comment by pantongren** [comment]
@Fleker ,I tried "turn on light" on web simulator and it 
correctly responded. But  when I closed and reopened the web simulator to test it ,it didn't work, just replied"I'm sorry, there was an error and I'm unable to control your home device", and it would continue this reply unless I unlink and relink my Accont. why? Is there anything expired after I close the web simulator?

**Comment by Fleker** [comment]
@pantongren it may be an OAuth issue that happens when you start a new session.

**Comment by pantongren** [comment]
@Fleker  thanks, 
Is there any solution?

**Comment by Fleker** [comment]
@pantongren not sure what a good solution may be, but this sounds like something you should put into a separate issue. Are you only using the web simulator? How are you doing OAuth? Detail your answers in that new thread.

**Comment by pantongren** [comment]
@Fleker   I tried to reset my Google Home Device a few days ago,but  failed to set it up during the last step of configuration. So, I can only use the web simulator or Allo App. I  have tried "turn on light" with Allo App, it replied "I can't quite control that yet". So, it works only by the web simulator.  I'm just testing this given sample which contains the whole OAuth process. I have done no modifacation with the sample OAuth,thanks.

**Comment by Fleker** [comment]
@pantongren please place further discussion of your issue in #22 as it's unrelated to this.

**Comment by xj13567180468** [comment]
my smart home  have the same issue   after token exchange. When trying to add devices with my test application,
I do get the error message that the "Cloudn't update setting. Check your connection".I tried many times.why?who can help me.

**Comment by xj13567180468** [comment]
my smart home have the same issue after token exchange. When trying to add devices with my test application,
I do get the error message that the "Cloudn't update setting. Check your connection".I tried many times.why?who can help me. @pantongren @Fleker @marziman @bartvollebregt @mwidlok

**Comment by Fleker** [comment]
@xj13567180468 you can email your project id and email [to me](mailto:fleker@google.com).

**Comment by Fleker** [comment]
We just pushed out a server-side update to the Smart Home servers. I have successfully SYNC'd my devices with a test account that was affected. Can others check and confirm that they are unblocked?

**Comment by xj13567180468** [comment]
we had email to you project id.

**Comment by xj13567180468** [comment]
@Fleker we are unable to SYNC test devices

**Comment by BenjaminBertran** [comment]
@Fleker It now works for me!! Thx

**Comment by Fleker** [comment]
I am closing this issue since it appears like the server fix has unblocked developers from getting a SYNC command through the Home app, allowing them to correctly execute commands. Thank you for your patience, and I'm excited to see what you make.

**Comment by patrickgao0922** [comment]
We just try to Google application tried to link accounts and SYNC however, the problem is still there and Route for SYNC didn't get triggered.

**Comment by xj13567180468** [comment]
the problem is still there and Route for SYNC didn't get triggered.we still had this problem。

**Comment by marziman** [comment]
@Fleker it worked for me and I was able to SYNC and EXEC.
Thanks to you and the team..I am sure you guys had alot stress with this topic.

BR Mehmet

**Comment by sunjunlishi** [comment]
i  still have the  smae error ,why?

**Comment by sunjunlishi** [comment]
dear @marziman ,how did you resolve it, i am very Ardent desire your advice.

**Comment by sunjunlishi** [comment]
my project id is  dev-xingtel

**Comment by sunjunlishi** [comment]
authorization code type author have got it,thanks, all my accounts is ok.

**Comment by bzhang222** [comment]
I have my own OAuth2 server which worked for Alexa, but I still have this issue with Google, what is the solution exactly?  are we saying it only works for google's sample OAuth2 implementation?

**Comment by sunjunlishi** [comment]
authorization code type author have got it,is your implicit autho type?

**Comment by madgeni** [comment]
Hi - just wanted to say this isn't working for me either - can see `11-06 19:07:38.171  7315 11191 I ActivityManager: START u0 {dat=assistant-handoff://complete/?result_code=SUCCESS&result_message=Accounts+now+linked. cmp=com.google.android.googlequicksearchbox/com.google.android.apps.gsa.assistant.handoff.BrowserControlActivity} from uid 10024 pid 1184 on display 0` - but seeing same behaviour as @dbrammertz

**Comment by Fleker** [comment]
Let's use issue #58 to continue discussing the issue

**Comment by ash-lat** [comment]
Hi
Its not working for me either, looks like it is not able to sync to /smarthome
keep getting error "POST /smarthome 500 9.289 ms - 1456"
I did enter https://xxxxxxxx.ngrok.io/smarthome to webhook URL in "Smart Home Action" panel
And on google assistant app it links succesfully but throws an error "Something went wrong. Try again"

Attached is log file
[gaerror.txt](https://github.com/actions-on-google/actionssdk-smart-home-nodejs/files/1713215/gaerror.txt)

Just wanted to confirm :- We have to use same client id and secret id as mentioned in config-provider.js? and only change the API key correct?
